atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -2378,6 +2364,37 @@ private boolean rollbackOtherTransaction(long startTs, TransactionService transa
         }
     }
 
+    ///////////////////////////////////////////////////////////////////////////
+    /// Post-commit actions
+    ///////////////////////////////////////////////////////////////////////////
+
+    private void runPostCommitActions() {
+        // Run close() to release locks before running success callbacks, since success callbacks might
+        // start a new transaction and attempt to grab the same locks as the current transaction.
+        close();
```

Comment:
Is there a world where this throws, but the transaction has actually committed, THEREFORE you want to make sure the callbacks run?

Reply:
Modified to `try-finally` to guarantee success callbacks run if transaction succeeded.